src/include/server/mir/scene/idle_hub.h

```diff
@@ -50,6 +50,16 @@ class IdleHub
     IdleHub() = default;
     virtual ~IdleHub() = default;
 
+    /// Lock that is held by the surface which is inhibiting idle
+    struct WakeLock
+    {
+    public:
+        WakeLock() = default;
+        virtual ~WakeLock() = default;
+        WakeLock(WakeLock const&) = delete;
+        WakeLock& operator=(WakeLock const&) = delete;
+    };
```

Comment:
There's no need for this type to be **defined** in the header. (Only declared)

If the definition is moved to the .cpp, then you can also refactor by "pushing up" the derived class implementation and lose the derived class.

Reply:
I am a bit confused: are you suggesting that WakeLock be defined within `idle_hub.cpp`? If so, that doesn't exist. 

If instead it should be defined within `basic_idle_hub.cpp`, should WakeLock still be declared within `idle_hub.h`? I don't believe non-defined structs can be inherited as they're incomplete types.

Reply:
Yes, declare `WakeLock` in `idle_hub.h`, just remove the definition from there and define in `basic_idle_hub.cpp` - the only place a definition is used or needed.

Reply:
I am having an issue where I cannot inherit `IdleHub::WakeLock` when it is undefined in `idle_hub.h` as it is considered an incomplete type.

Maybe you can help clear up some confusion surrounding inheritance that I've been struggling with:
If we are declaring that `IdleHub::inhibit_idle() -> std::shared_ptr<WakeLock>`, but we withhold the implementation details of the WakeLock until `BasicIdleHub::WakeLock`, how can the incomplete type `IdleHub::WakeLock` be held in `BasicIdleHub::wake_lock` which is a `std::weak_ptr<IdleHub::WakeLock>`? 

How can the destructor that we are declaring inside `BasicIdleHub::WakeLock` be called if its type is treated as `IdleHub::WakeLock` and not `BasicIdleHub::WakeLock`?

If `BasicIdleHub::wake_lock` should hold type `std::weak_ptr<BasicIdleHub::WakeLock>`, then `BasicIdleHub::inhibit_idle()` can not match the interface and return an `std::shared_ptr<IdleHub::WakeLock>` as that would require downcasting to an incomplete type.

Reply:
1. Why do you want to introduced a `BasicIdleHub::WakeLock`?
2. `std::weak_ptr<>` can hold an incomplete type.
3. If you use `make_shared<>` then it creates a _deleter_ that deletes the creation type.

Here's what I mean:

```
diff --git a/src/include/server/mir/scene/idle_hub.h b/src/include/server/mir/scene/idle_hub.h
index 5f4ff073a7..65a1df5194 100644
--- a/src/include/server/mir/scene/idle_hub.h
+++ b/src/include/server/mir/scene/idle_hub.h
@@ -51,14 +51,7 @@ public:
     virtual ~IdleHub() = default;
 
     /// Lock that is held by the surface which is inhibiting idle
-    struct WakeLock
-    {
-    public:
-        WakeLock() = default;
-        virtual ~WakeLock() = default;
-        WakeLock(WakeLock const&) = delete;
-        WakeLock& operator=(WakeLock const&) = delete;
-    };
+    struct WakeLock;
 
     /// Wakes Mir if it's not already awake, and starts the timer from now
     virtual void poke() = 0;
diff --git a/src/server/scene/basic_idle_hub.cpp b/src/server/scene/basic_idle_hub.cpp
index 2abece0744..bffe8cfb81 100644
--- a/src/server/scene/basic_idle_hub.cpp
+++ b/src/server/scene/basic_idle_hub.cpp
@@ -265,26 +265,27 @@ void ms::BasicIdleHub::schedule_alarm(ProofOfMutexLock const&, time::Timestamp c
     }
 }
 
-auto ms::BasicIdleHub::inhibit_idle() -> std::shared_ptr<WakeLock>
+/// Lock that is held by the surface which is inhibiting idle
+struct ms::IdleHub::WakeLock
 {
-    struct WakeLock: public IdleHub::WakeLock
+    WakeLock(std::weak_ptr<IdleHub> idle_hub) : idle_hub{std::move(idle_hub)}
     {
-        WakeLock(std::weak_ptr<IdleHub> idle_hub) : idle_hub{std::move(idle_hub)}
-        {
-        }
+    }
 
-        ~WakeLock()
+    ~WakeLock()
+    {
+        if (auto const shared_hub = idle_hub.lock())
         {
-            if (auto const shared_hub = idle_hub.lock())
-            {
-                shared_hub->poke();
-            }
+            shared_hub->poke();
         }
+    }
 
-    private:
-        std::weak_ptr<IdleHub> const idle_hub;
-    };
+private:
+    std::weak_ptr<IdleHub> const idle_hub;
+};
 
+auto ms::BasicIdleHub::inhibit_idle() -> std::shared_ptr<WakeLock>
+{
     std::lock_guard<std::mutex> lock(mutex);
     if (auto const shared_wake_lock = wake_lock.lock()) // wake_lock is already held
     {
```

Reply:
> 1. Why do you want to introduced a `BasicIdleHub::WakeLock`?

This was my issue. I was introducing a new type of WakeLock that would inherit from `idle_hub.h`.